### PR TITLE
New Solver: Iterative Hessian Sketch

### PR DESCRIPTION
**Pull request by VChristian**
Implemented iterate hessian sketching as another method to solve linear systems. This branch includes:

- New mutable struct `IterativeHessianSketch` in `src/linear_solver_routines/iterative_hessian_sketch.jl`
- `rsubsolve!` interface for `IterativeHessianSketch` which implements the algorithm.
- Testing file for `IterativeHessianSketch` and related `rsubsolve!` procedure.

### CODE REVIEW

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    ##########################
+    # check final solutions
+    rsub = IterativeHessianSketch(A,b,nothing,nothing)
+    x0 = zeros(5)
+    for i in 1:100
+        S = randn(5*10,length(b))
+        RLinearAlgebra.rsubsolve!(rsub, x0, (S, S*A, S*A*x0 - S*b), i)
+    end
+    ldivsol = zeros(5)
+    LinearAlgebra.ldiv!(ldivsol,qr(A'*A),A'*b)
```
[the remark this answers is not in the document]
Comment (npritch928) [reply test/linear_solver_routines/proc_iterative_hessian_sketch.jl:100]: Add space around `*`

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    ##########################
+    ##########################
+
+    ########################## Sketch size >= number of columns
+    ##########################
+    # check final solutions
+    rsub = IterativeHessianSketch(A,b,nothing,nothing)
+    x0 = zeros(5)
+    for i in 1:100
+        S = randn(5*10,length(b))
```
[the remark this answers is not in the document]
Comment (npritch928) [reply test/linear_solver_routines/proc_iterative_hessian_sketch.jl:96]: Why not just have `randn(50, length(b))`

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    # User inputed values should be rewritted on the first iteration
+    @test length(rsub.btilde) == 5
+    @test length(rsub.step) == 5
+
+    # test to make sure rsub.btilde is correctly initialized.
+    @test size(S)[1]*A'*(b-A*zeros(5)) ≈ rsub.btilde 
+    
+    # check to make sure that the inner problem is solved correctly (correct step)
+    _,R=qr(S*A)
+    @test norm(R'*R*rsub.step - size(S)[1]*A'*(b-A*zeros(5)) ) < 1e-10
```
[the remark this answers is not in the document]
Comment (npritch928) [reply test/linear_solver_routines/proc_iterative_hessian_sketch.jl:83]: Add spaces around `*`

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    # User inputed values should be rewritted on the first iteration
+    @test length(rsub.btilde) == 5
+    @test length(rsub.step) == 5
+
+    # test to make sure rsub.btilde is correctly initialized.
+    @test size(S)[1]*A'*(b-A*zeros(5)) ≈ rsub.btilde 
+    
+    # check to make sure that the inner problem is solved correctly (correct step)
+    _,R=qr(S*A)
```
[the remark this answers is not in the document]
Comment (npritch928) [reply test/linear_solver_routines/proc_iterative_hessian_sketch.jl:82]: Add space around `*`

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    # test user inputed values
+    @test size(rsub.btilde)[1] == 123
+    @test size(rsub.step)[1] == 123 
+    @test rsub.btilde == btilde_random
+    @test rsub.step == step_random
+
+    # one step of the algorithm
+    x0 = zeros(5)
+    S = randn(5,length(b))
+    RLinearAlgebra.rsubsolve!(rsub, x0, (S, S*A, S*A*x0 - S*b), 1)
```
Comment (npritch928) [inline test/linear_solver_routines/proc_iterative_hessian_sketch.jl:72]: Add space around `*`

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    # check to make sure btilde and step get initialized to the correct length
+    @test length(rsub.btilde) == 5
+    @test length(rsub.step) == 5
+
+    # test to make sure rsub.btilde is correctly initialized.
+    @test size(S)[1]*A'*(b-A*zeros(5)) ≈ rsub.btilde
+
+    # check to make sure that the inner problem is solved correctly (correct step)
+    _,R=qr(S*A)
+    @test norm(R'*R*rsub.step - size(S)[1]*A'*(b-A*zeros(5)) ) < 1e-10
```
[the remark this answers is not in the document]
Comment (npritch928) [reply test/linear_solver_routines/proc_iterative_hessian_sketch.jl:50]: Add spaces around `*`s.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    # form sub-linear system and solve 
+    type.btilde .= m .* ((type.A)'*(type.b - type.A*x))
+    if m >= size(samp[2])[2] # nrow >= ncol for SA
+        _,R = qr(samp[2])
+        type.step .= R'\type.btilde
+        type.step .= R\type.step
+    else
+        try
+            LinearAlgebra.ldiv!(type.step, qr(samp[2]'*samp[2]), type.btilde)
```
Comment (npritch928) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:61]: Add spaces around `*`.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        if m < d 
+            @warn "Sketch matrix might be too small for a sensible inner problem solution!"
+        end
+    end
+
+    # form sub-linear system and solve 
+    type.btilde .= m .* ((type.A)'*(type.b - type.A*x))
+    if m >= size(samp[2])[2] # nrow >= ncol for SA
+        _,R = qr(samp[2])
+        type.step .= R'\type.btilde
```
Comment (npritch928) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:57]: Add spaces around `\`

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        type.btilde = Array{typeof(samp[2][1])}(undef, d)
+
+        # sketch matrix will not have full column rank
+        if m < d 
+            @warn "Sketch matrix might be too small for a sensible inner problem solution!"
+        end
+    end
+
+    # form sub-linear system and solve 
+    type.btilde .= m .* ((type.A)'*(type.b - type.A*x))
```
[the remark this answers is not in the document]
Comment (npritch928) [reply src/linear_solver_routines/iterative_hessian_sketch.jl:54]: Add spaces around `*`s

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            LinearAlgebra.ldiv!(type.step, qr(samp[2]'*samp[2]), type.btilde)
+        catch
+            type.step .= zeros(size(x)[1])
+            @warn "Sketch size might be too small. Encountered error in LinearAlgebra.ldiv!, no update applied!"
+        end
+    end
+
+    # update current iterate
+    x .= x .+ type.step
+
```
Comment (npritch928) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:70]: Remove this extra white space.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    if m >= size(samp[2])[2] # nrow >= ncol for SA
+        _,R = qr(samp[2])
+        type.step .= R'\type.btilde
+        type.step .= R\type.step
+    else
+        try
+            LinearAlgebra.ldiv!(type.step, qr(samp[2]'*samp[2]), type.btilde)
+        catch
+            type.step .= zeros(size(x)[1])
+            @warn "Sketch size might be too small. Encountered error in LinearAlgebra.ldiv!, no update applied!"
```
Comment (npritch928) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:64]: As was mentioned early this warning could be made even more clear by replacing "Sketch size" with "The sampler's blockSize".

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    @test rsub.b == b
+
+    # Verify that buffer arrays are nothing
+    @test isnothing(rsub.step)
+    @test isnothing(rsub.btilde)
+
+    # one step of the algorithm
+    x0 = zeros(5)
+    S = randn(5,length(b))
+    RLinearAlgebra.rsubsolve!(rsub, x0, (S, S*A, S*A*x0 - S*b), 1)
```
Comment (npritch928) [inline test/linear_solver_routines/proc_iterative_hessian_sketch.jl:39]: Add spaces around `*`s.

File: test/linear_solver_routines/proc_iterative_hessian_sketch.jl [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+    # check to make sure btilde and step get initialized to the correct length
+    @test length(rsub.btilde) == 5
+    @test length(rsub.step) == 5
+
+    # test to make sure rsub.btilde is correctly initialized.
+    @test size(S)[1]*A'*(b-A*zeros(5)) ≈ rsub.btilde
```
Comment (npritch928) [inline test/linear_solver_routines/proc_iterative_hessian_sketch.jl:46]: Add spaces around `*`s and `-`s.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+See Mert Pilanci and Martin J. Wainwright. "Iterative Hessian Sketch: 
+    Fast and Accurate Solution Approximation for Constrained Least-Squares."
+    Journal of Machine Learning Research, 17(53), 1–38.
+    http://jmlr.org/papers/v17/14-460.html
+
+# Fields
+
+- `A::AbstractArray` is the coefficient matrix for the linear system. Required as we need to compute the full residual.
+- `b::AbstractVector` is the constant term (i.e., the linear system is Ax=b).
```
Comment (npritch928) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:18]: If you are noting that the matrix `A` is required to compute the residual the same note should be made for the constant vector.

**Review by npritch928 — CHANGES_REQUESTED** [review]
I think overall the code is really good and really like your care with the warnings and error messages. I have made a few suggestions on implementation details and have included a bunch of small changes so that the formatting matches the rest of the package.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+A mutable structure that represents the Iterative Hessian Sketch Method.
+
+See Mert Pilanci and Martin J. Wainwright. "Iterative Hessian Sketch: 
+    Fast and Accurate Solution Approximation for Constrained Least-Squares."
+    Journal of Machine Learning Research, 17(53), 1–38.
+    http://jmlr.org/papers/v17/14-460.html
+
+# Fields
+
+- `A::AbstractArray` is the coefficient matrix for the linear system. Required as we need to compute the full residual.
```
[the remark this answers is not in the document]
Comment (VChristian) [reply src/linear_solver_routines/iterative_hessian_sketch.jl:17]: I will fix that! In `vec_col_projection_fortho.jl` and `vec_row_projection_fortho.jl`, the comma is not there. I can fix that on my branch?

File: src/linear_solver_routines/iterative_hessian_sketch.jl (hunk 7 above)
Comment (VChristian) [reply src/linear_solver_routines/iterative_hessian_sketch.jl:61]: How come it is better to do a `cholesky`?

File: src/linear_solver_routines/iterative_hessian_sketch.jl (hunk 10 above)
Comment (VChristian) [reply src/linear_solver_routines/iterative_hessian_sketch.jl:70]: Which extra white space? Do you mean the new line inbetween `x .= x .+ type.step` and `return nothing`?

Comment (npritch928) [reply src/linear_solver_routines/iterative_hessian_sketch.jl:78]: The empty line on line 78 is unnecessary. Sorry for not being clear.

**Review by npritch928 — APPROVED** [review]
Great Job!

File: src/RLinearAlgebra.jl [hunk 16]
```diff
@@ -60,7 +60,7 @@ export LinSysSolveRoutine, LinSysVecRowProjection, LinSysVecColProjection,
     LinSysBlkRowProjection, LinSysBlkColProjection, LinSysPreconKrylov
 
 # Vector Row Projection
-export LinSysVecRowProjStd, Kaczmarz, ART, LinSysVecRowProjPO, LinSysVecRowProjFO
+export LinSysVecRowProjStd, Kaczmarz, ART, LinSysVecRowProjPO, LinSysVecRowProjFO, IterativeHessianSketch
```
Comment (vp314) [inline src/RLinearAlgebra.jl:63]: Move under block row projection

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+See Mert Pilanci and Martin J. Wainwright. "Iterative Hessian Sketch: 
+    Fast and Accurate Solution Approximation for Constrained Least-Squares."
+    Journal of Machine Learning Research, 17(53), 1–38.
+    http://jmlr.org/papers/v17/14-460.html
+
+# Fields
+
+- `A::AbstractArray`, is the coefficient matrix for the linear system. Required as we need to compute the full residual.
+- `b::AbstractVector`, is the constant term (i.e., the linear system is Ax=b). Required as we need to compute the full residual.
```
Comment (vp314) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:18]: Remove `is`

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    # initialize buffer arrays
+    m = size(samp[1])[1] # sketch size
+    if iter == 1
+        d = size(x)[1]
+        type.step = Array{typeof(samp[2][1])}(undef, d) 
+        type.btilde = Array{typeof(samp[2][1])}(undef, d)
+
+        # sketch matrix will not have full column rank
+        if m < d 
+            @warn "The sampler's blockSize might be too small for a sensible inner problem solution!"
```
Comment (vp314) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:53]: Why is `blockSize` one word?

Comment (vp314) [reply src/linear_solver_routines/iterative_hessian_sketch.jl:53]: Remove the exclamation point.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    if m >= size(samp[2])[2]
+        R = qr(samp[2]).R
+        type.step .= R' \ type.btilde
+        type.step .= R \ type.step
+    else # if QR cannot be done, form full sketched Hessian and try to solve
+        try
+            LinearAlgebra.ldiv!(type.step, cholesky(samp[2]' * samp[2]), type.btilde)
+        catch
+            type.step .= zeros(size(x)[1])
+            @warn "The sampler's blockSize might be too small. Encountered error in LinearAlgebra.ldiv!, no update applied!"
```
Comment (vp314) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:71]: Why is `blockSize` one word? Remove exclamation point at the end.

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+        end
```
Comment (vp314) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:72]: Use backsolve here. Needs a warning that if `m <= n` there is no theory supporting this solver.

**Review by vp314 — CHANGES_REQUESTED** [review]

File: src/linear_solver_routines/iterative_hessian_sketch.jl [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    Fast and Accurate Solution Approximation for Constrained Least-Squares."
+    Journal of Machine Learning Research, 17(53), 1–38.
+    http://jmlr.org/papers/v17/14-460.html
+
+# Fields
+
+- `A::AbstractArray`, the coefficient matrix for the linear system. Required as we need to compute the full residual.
+- `b::AbstractVector`, the constant term (i.e., the linear system is Ax=b). Required as we need to compute the full residual.
+- `step::Union{AbstractVector, Nothing}`, buffer array to hold solution of subproblem that provides update to x.
+- `btilde::Union{AbstractVector, Nothing}`, buffer array to hold constant term in forming the subproblem to compute `step`.
```
Comment (vp314) [inline src/linear_solver_routines/iterative_hessian_sketch.jl:20]: Need information on constructors

**Review by vp314 — CHANGES_REQUESTED** [review]

**Review by vp314 — APPROVED** [review]

**Pull request merged**